plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+
+
+float SaProcessor::binBandwidth()
+{
+	return (getSampleRate() / 2.0) / binCount();
+}
+
+
+float SaProcessor::getFreqRangeMin(bool linear)
+{
+	switch (m_controls->m_freqRangeModel.value())
+	{
+		case FRANGE_AUDIBLE: return FRANGE_AUDIBLE_START;
+		case FRANGE_BASS: return FRANGE_BASS_START;
+		case FRANGE_MIDS: return FRANGE_MIDS_START;
+		case FRANGE_HIGH: return FRANGE_HIGH_START;
+		default:
+		case FRANGE_FULL: return linear ? 0 : LOWEST_LOG_FREQ;
+	}
+}
+
+
+float SaProcessor::getFreqRangeMax()
+{
+	switch (m_controls->m_freqRangeModel.value())
+	{
+		case FRANGE_AUDIBLE: return FRANGE_AUDIBLE_END;
+		case FRANGE_BASS: return FRANGE_BASS_END;
+		case FRANGE_MIDS: return FRANGE_MIDS_END;
+		case FRANGE_HIGH: return FRANGE_HIGH_END;
+		default:
+		case FRANGE_FULL: return getSampleRate() / 2;
+	}
+}
+
+
+// Map frequency to pixel x position on a display of given width.
+float SaProcessor::freqToXPixel(float freq, int width)
+{
+	if (m_controls->m_logXModel.value())
+	{
+		if (freq <= 1) {return 0;}
+		float min = log10(getFreqRangeMin());
+		float range = log10(getFreqRangeMax()) - min;
+		return (log10(freq) - min) / range * width;
+	}
+	else
+	{
+		float min = getFreqRangeMin();
+		float range = getFreqRangeMax() - min;
+		return (freq - min) / range * width;
+	}
+}
+
+
+// Map pixel x position on display of given width back to frequency.
+float SaProcessor::xPixelToFreq(float x, int width)
+{
+	if (m_controls->m_logXModel.value())
+	{
+		float min = log10(getFreqRangeMin());
+		float max = log10(getFreqRangeMax());
+		float range = max - min;
+		return pow(10, min + x / width * range);
+	}
+	else
+	{
+		float min = getFreqRangeMin();
+		float range = getFreqRangeMax() - min;
+		return min + x / width * range;
+	}
+}
+
+
+// --------------------------------------
+// Amplitude conversion helpers
+//
+float SaProcessor::getAmpRangeMin(bool linear)
+{
+	// return very low limit to make sure zero gets included at linear grid
+	if (linear) {return -900;}
+	switch (m_controls->m_ampRangeModel.value())
+	{
+		case ARANGE_EXTENDED: return ARANGE_EXTENDED_START;
+		case ARANGE_AUDIBLE: return ARANGE_AUDIBLE_START;
+		case ARANGE_NOISE: return ARANGE_NOISE_START;
+		default:
+		case ARANGE_DEFAULT: return ARANGE_DEFAULT_START;
+	}
+}
+
+
+float SaProcessor::getAmpRangeMax()
+{
+	switch (m_controls->m_ampRangeModel.value())
+	{
+		case ARANGE_EXTENDED: return ARANGE_EXTENDED_END;
+		case ARANGE_AUDIBLE: return ARANGE_AUDIBLE_END;
+		case ARANGE_NOISE: return ARANGE_NOISE_END;
+		default:
+		case ARANGE_DEFAULT: return ARANGE_DEFAULT_END;
+	}
+}
+
+
+// Map amplitude to pixel y position on a display of given height.
+float SaProcessor::ampToYPixel(float amplitude, int height)
+{
+	if (m_controls->m_logYModel.value())
+	{
+		if (10 * log10(amplitude) < getAmpRangeMin())
+		{
+			return height;
+		}
+		else
+		{
+			float max = getAmpRangeMax();
+			float range = getAmpRangeMin() - max;
+			return (10 * log10(amplitude) - max) / range * height;
```

Comment:
`10 *` is the "deci-" part in "decibel". :)

Reply:
So I think I got this now. By definition of `bel`, the result of `log10` is bels, and
```
1 bel = 10 * 1/10 bel = 10 * 1 dB
```
which explains the `10*`. If you think it's not obvious, can you please add a comment for this code line? Though, if it's obvious enough, don't.

Reply:
Yeah, I got it backwards there; the formula / conversion you wrote is correct.

If I describe the `10 *` part, it kind-of wouldn't have much sense on its own. I think mentioning dB in the comments should be enough, because if you don't understand decibels, you will have to look it up anyway. And given all the messy details around it, I'm not sure if comments are the right place to explain it.
I'll try to fit in some comments mentioning a conversion to / from dB is taking place, but I won't go into details.